File: Code/C++/Check_Linked_List_Palindrome.cpp
/*
In this code we take values of linked list nodes from user and then call the function "check_palindrome",
to check if given linked list is palindrome or not, by reversing the last half of the linked list and
then comparing it with the first half elements of the linked list.
*/

#include<iostream>
using namespace std;

struct Node{
	int val;
	Node* next;
};

Node* newNode(int n){
	Node* t = new Node;
	t->val = n;
	t->next = NULL;
}

	//this function reverse the given linked list
Node* reverse(Node* N){		
	Node* prev = NULL;
	Node* curr = N;
	Node* right = NULL;
	while(curr!=NULL){
		right = curr->next;
		curr->next = prev;
		prev = curr;
		curr = right;
	}
	return prev;
}

	//this function find the middle element of linked list
Node* middle(Node* N, int n){		
	Node* ans = N;
	for(int i=0;i<n/2;i++){
		ans = ans->next;
	}
	return ans;
}

	//this function checks if given linked list is palindrome or not
bool check_palindrome(Node* N, int n){		
	Node* v = middle(N,n);
	Node* M = reverse(v);
	for(int i=0;i<n/2;i++){
		if(M->val!=N->val){
			return false;
		}
	}
	return true;
}

int main(){
	cout<<"Enter number of Nodes: ";
	int n;
	cin>>n;
	Node* h = new Node;
	Node* head = h;
	
	//this loop convert the input given by user into a linked list
	for(int i=0;i<n;i++){		 
		int a;
		cin>>a;
		h->next = newNode(a);
		h = h->next;
	}
	head = head->next;
	bool p = check_palindrome(head,n);
	
	if(p==true){
		cout<<"Given Linked list is palindrome!";
	}
	else{
		cout<<"Given Linked list is not palindrome!";
	}
	
	return 0;
}

/*
Time Complexity = O(n);
Space Complexity = O(1);
 
If we give following linked list as input - 
3->5->6->8->6->5->3
Output - "Given Linked list is palindrome!"

If we give following linked list as input - 
3->5->6->8->7->5->3
Output - "Given Linked list is not palindrome!"
*/

File: Code/Java/Check_LinkedList_palindrome.java
/*    Check if the Linked List is palindrome or not.
      Approach: We will traverse the linkedlist to first find the middle element and then reverse the 
      second half of the linked list in O(n/2) time. 
      Then we will check from both the directions of the linked list to see if it is a palindrome or not.
      We print "Yes" if its a palindrome otherwise "No".
*/
import java.util.*;

public class Check_LinkedList_palindrome {

	public static class ListNode {
		      int val;
		      ListNode next;
		      ListNode() {}
		      ListNode(int val) { this.val = val; }
		      ListNode(int val, ListNode next) { this.val = val; this.next = next; }
		 }
	public static void main(String[] args) {
		Scanner sc= new Scanner(System.in);
		
		int n= sc.nextInt();
		ListNode head= new ListNode(sc.nextInt());
		ListNode tempp= head;
		for(int i=0; i<n-1; i++) {
			tempp.next= new ListNode(sc.nextInt());
		}
		
		ListNode temp= head;
        int count=0;
        while(temp!=null){
            temp=temp.next; count++;
        }
        count= count/2;
        // the middle number i.e. count/2 is calculated above.
    
        ListNode midnode= head;
        ListNode fast= head;
        while(fast.next!=null && fast.next.next!=null){
            midnode= midnode.next;
            fast= fast.next.next;
        }
        
        ListNode prev= midnode;
        midnode= midnode.next;
        
        //we will start reversing other half here
        ListNode curr= midnode;
        ListNode next= null;
        while(curr!=null){
            next= curr.next;
            curr.next= prev;
            prev= curr;
            curr= next;
        }
        ListNode lastNode= prev;
        
        while(count-->0){
            if(head.val!=lastNode.val) {
            	System.out.println("No");
            	System.exit(1);
            }
            head= head.next;
            lastNode= lastNode.next;
        }
        System.out.println("Yes");
        
	}

}

/*
INPUT:
5
1 2 3 4 5
OUTPUT:
No

INPUT:
6
1 2 3 3 2 1
OUTPUT:
Yes


Complexity Analysis:  
    * Time complexity: O(N)
      Only traversal of the loop till N is needed.
    * Auxiliary Space: O(1) 
      There is no space required.
*/
